Remove flush option from savers and removers, use base_options_resolver everywhere

iulyanp opened the pull request:
Hi @nidup ,

I'm back with the second PR where I removed the "flush", "schedule" and "recalculate" options from saver and remover classes.
Please take a look and let me know if you find anything that I missed or broke.
The only saver I didn't changed is GroupSaver, I'll update that one also if what I did here is fine with you.
In Pim\Bundle\CatalogBundle\EventSubscriber\ProductCategorySubscriber we can remove now the "flush" option from line 58: $this->saver->saveAll($productsToUpdate, ['flush'=> $event->getArgument('flush')]);, because the flush is always true.

After this changes some Savers (ex. FamilySaver) don't have any options and the resolve options could be removed entirely. What do you think?

Regards,
Iulian

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | N
| Changelog updated                 | -
| Review and 2 GTM                  | -
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

nidup commented on src/Pim/Bundle/CatalogBundle/Resources/config/savers.yml:
```diff
@@ -1,8 +1,8 @@
-    pim_catalog.saver.product_options_resolver.class:      Pim\Bundle\CatalogBundle\Doctrine\Common\Saver\ProductSavingOptionsResolver
-    pim_catalog.saver.completeness_options_resolver.class: Pim\Bundle\CatalogBundle\Doctrine\Common\Saver\CompletenessSavingOptionsResolver
+#    pim_catalog.saver.product_options_resolver.class:      Pim\Bundle\CatalogBundle\Doctrine\Common\Saver\ProductSavingOptionsResolver
+#    pim_catalog.saver.completeness_options_resolver.class: Pim\Bundle\CatalogBundle\Doctrine\Common\Saver\CompletenessSavingOptionsResolver
```
You can drop the 2 classes and the related ProductSavingOptionsResolverSpec and CompletenessSavingOptionsResolverSpec

nidup commented on src/Pim/Bundle/CatalogBundle/Resources/config/savers.yml:
```diff
@@ -110,14 +110,14 @@ services:
-    pim_catalog.saver.completeness_options_resolver:
-        class: '%pim_catalog.saver.completeness_options_resolver.class%'
+#    pim_catalog.saver.completeness_options_resolver:
+#        class: '%pim_catalog.saver.completeness_options_resolver.class%'
```
ok to remove these services

nidup commented:
Hi @iulyanp 

Thank you very much for the pull request! I'm doing the related EE PR in the same time :)

You can re-work the GroupSaver in the same way.

Ok to remove the flush option in ProductCategorySubscriber.

I discussed with @jjanvier and we've seen something weird, we should always dispatch the options in the event, like this "new GenericEvent($products, $options));". We do it sometimes, we should do it in any saver and remover to be consistent, thank you to fix this.

We also discussed about the saving and removing options resolver and in fact it's strange to expose a such internal behavior as a dedicated service. So you can drop all BaseSavingOptionsResolver, ProductSavingOptionResolver, BaseRemovingOptionsResolver, etc. If a Saver or a Remover needs an option resolver, like the GroupSaver, we can instanciate it in the Saver or the Remover directly (in a protected method).

Could you fix the php specs too by running "bin/phpspec run" ? If you're not comfortable with phpspec let me know and i'll fix them.

Don't hesitate to ping me if you need anything, I'll be glad to help.

PS : we'll release the v1.6.0 the 30th so we'll have to merge these PRs this week or next week, if we can't merge, no pb, we'll merge in the v1.7.

iulyanp commented:
Hello @nidup 

I'm happy to inform you that I finally finished refactoring the 'saver & remover' classes. :fireworks: 
Could you please review my PR and let me know if there is anything that I missed.
I removed the services for Option Resolvers that we don't use and deleted the classes and spec tests.
The only option resolver I didn't delete is GroupOptionsResolver, I just removed the flush option and worked the save and saveAll functions there.
In VariantGroupController I deleted the flush option and I noticed some small mistake with "$id" that wasn't defined and I changed it with $code in the exception messages.
Also I fixed the php spec tests and I think everything should work accordingly.

Best regards,
Iulian

nidup commented:
Hi @iulyanp 

Thanks a lot for this contribution, I reviewed and didn't see any issue, it really makes the whole system more clean and understandable! :tada: 

I runned all our behat builds on ORM and MongoDB storages and all are tests are passing without any problems :tada:

I'm re-working the Enterprise Edition to make it compatible with these changes, i'm doing good progress, I think it will be done at the beginning of the next week, I keep you posted once done.

Thanks again it's a real pleasure to work with you!

iulyanp commented:
Hi @nidup 

Thanks again for the quick review, I'm glad that everything looks good. 
Can't wait to see v1.6 released. :)

Regards,

nidup commented:
Hi @iulyanp !

:tada: I just finished to update the enterprise edition, all builds are ok and I merged both pull requests, this improvement will be released in v1.6 :tada: 

Thanks again for this contribution from the whole Akeneo core team, it's very valuable for us!

FYI, until now, we managed our whole roadmap in our internal ticketing system for stories, tech improvements and bugs. We're now trying to publish more tech improvs and bugs as github issues to allow to everybody to contribute on community edition, and especially on bigger pull requests as the one you did.

To be able to list the issues on which we'd love to be helped, we introduced a new github label https://github.com/akeneo/pim-community-dev/issues?q=is%3Aissue+is%3Aopen+label%3A%22wanna+contribute%3F%22

Don't hesitate to propose improvements as new issues or to open pull request to fix existing issues.

Last but not least, we recently released Badger, a platform originally used to gamify our internal day to day work at Akeneo and which can now be used by the community too, if you're interested you can claim the following badge, you deserve it! http://badger.akeneo.com/badge/41acec2c-649f-11e6-92dc-d60437e930cf

Looking forward to work again with you, have a good weekend!
